src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs

```diff
@@ -476,7 +478,19 @@ private void ProcessEvents()
                         string expandedName = null;
                         WatchedDirectory associatedDirectoryEntry = null;
 
-                        if (nextEvent.wd != -1) // wd is -1 for events like IN_Q_OVERFLOW that aren't tied to a particular watch descriptor
+                        // Two events take priority: Overflow and SymLinks. An overflow event means that we can't trust 
+                        // our state without restarting since we missed events and some of those events could be a 
+                        // directory create, meaning we wouldn't have added the directory to the watch and would not provide
+                        // correct data to the caller. The other event, SymLinks, should be filtered out due to the FSW contract.
+                        if ((mask & (uint)Interop.libc.NotifyEvents.IN_Q_OVERFLOW) != 0)
+                        {
+                            // Notify the caller of the error and restart to pick up any potential directories we missed due
+                            // to the overflow.
+                            watcher.NotifyInternalBufferOverflowEvent();
+                            watcher.Restart();
```

Comment:
We only need to restart on Linux because on Windows and OSX we only watch the root directory and we are told automatically about the subdirectories; on Linux, we need to watch the subdirectories manually, so if we got an overflow then we need to restart in order to do a full enumeration of the file system and add watches on all subfolders. You brought up a good point in that we should only do this when the bit is set to watch subdirectories